### PR TITLE
WIP: Remove `CommitStatus.Index` field.

### LINKED ISSUE #14059: Commit build status not updating (intermittent)
- Gitea version (or commit ref): 1.14.0+dev-414-g4ef751215 
- Git version: 2.20.1
- Operating system: Debian based
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [x] MySQL
  - [ ] MSSQL
  - [ ] SQLite

## Description

Occasionally one of my Drone builds will not update it's status properly, leaving the commit as "Build in progress" with the yellow circle, when actually checking Drone it shows "Successful", or "Failed".

This only happens occasionally, the only pattern I can see is its only been for PR's so far, and also when the server is under heavy load. I first noticed when I ran "update branch" on 4 PR's at the same time, each calls 2 builds, so I have many builds queued.

From the logs, the error seems to be a SQL duplicate value in the unique index.

```
[17/Dec/2020:09:27:30 +0000] "GET /api/v1/repos/xxxx/xxxx/raw/a12437fc1f5b4153a3e25f9b8d82a39d20b7e7e8/.drone.yml HTTP/1.1" 200 2553 "-" "Go-http-client/1.1"
[17/Dec/2020:09:27:30 +0000] "POST /api/v1/repos/xxxx/xxxx/statuses/a12437fc1f5b4153a3e25f9b8d82a39d20b7e7e8 HTTP/1.1" 201 638 "-" "Go-http-client/1.1"
[17/Dec/2020:09:27:31 +0000] "POST /api/v1/repos/xxxx/xxxx/statuses/a12437fc1f5b4153a3e25f9b8d82a39d20b7e7e8 HTTP/1.1" 201 636 "-" "Go-http-client/1.1"
[17/Dec/2020:09:45:09 +0000] "POST /api/v1/repos/xxxx/xxxx/statuses/a12437fc1f5b4153a3e25f9b8d82a39d20b7e7e8 HTTP/1.1" 201 638 "-" "Go-http-client/1.1"
[17/Dec/2020:09:45:22 +0000] "POST /api/v1/repos/xxxx/xxxx/statuses/a12437fc1f5b4153a3e25f9b8d82a39d20b7e7e8 HTTP/1.1" 201 636 "-" "Go-http-client/1.1"
[17/Dec/2020:09:50:59 +0000] "POST /api/v1/repos/xxxx/xxxx/statuses/a12437fc1f5b4153a3e25f9b8d82a39d20b7e7e8 HTTP/1.1" 201 636 "-" "Go-http-client/1.1"
[17/Dec/2020:09:50:59 +0000] "POST /api/v1/repos/xxxx/xxxx/statuses/a12437fc1f5b4153a3e25f9b8d82a39d20b7e7e8 HTTP/1.1" 500 375 "-" "Go-http-client/1.1"
```

```
2020/12/17 09:50:59 ...pi/v1/repo/status.go:64:NewCommitStatus() [E] CreateCommitStatus: NewCommitStatus[repo_id: 1, user_id: 1, sha: a12437fc1f5b4153a3e25f9b8d82a39d
20b7e7e8]: Insert CommitStatus[/home/git/gitea-repositories/xxxx/xxxx.git, a12437fc1f5b4153a3e25f9b8d82a39d20b7e7e8]: Error 1062: Duplicate entry '5-1-a12437fc1f5b
4153a3e25f9b8d82a39d20b7e7e8' for key 'UQE_commit_status_repo_sha_index'
```

Looking at how the index is generated, I'm guessing two status updates came in simultaneously and both generated the same value? Maybe a lock on the db table is needed during index increment?

https://github.com/go-gitea/gitea/blob/4aabbacdfe2830e3b3808e42d6d89d6688754089/models/commit_status.go#L220-L240

~~I notice that file has been updated yesterday, but looking at the code it looks like the index is still generated the same way. I will update my gitea anyway and keep an eye on things.~~ Just happened again on latest master. I've updated Gitea version above.

### PR DESCRIPTION
**Pull request by bobemoe**
The problem is if two `commit status` updates come in simultaneously, they get the same `Index` value, and fail to insert into db, resulting in missing data. (described in and fixes #14059)

As far as I can tell this field is only used to know the sequence for commit statuses of one commit.  This can be calculated from the `ID` field, with no need for the additional `Index` field.

This PR removes the `Index` field and refactors the ordering to use the `ID` field.

I see no other use of `CommitStatus.Index`, apart from the API exposes ID which used to map to `Index` and I've changed this to map to `ID`.

## TODO
 - [x] Remove `Index` field
 - [x] Sort by `ID` instead of `Index`
 - [ ] Consider API impact
 - [ ] Fix tests
 - [ ] Migration
 - [ ] Testing

### CODE REVIEW

**Comment by bobemoe** [comment]
I don't understand the mysql8 test failure? Did something timeout?


```
...

=== TestSignupEmail (integrations/signup_test.go:37)
=== TestPushDeployKeyOnEmptyRepo (integrations/ssh_key_test.go:47)
SIGABRT: abort
PC=0x474061 m=0 sigcode=0

goroutine 0 [idle]:
runtime.futex(0x42a5b08, 0x80, 0x0, 0x0, 0x200000000, 0x42a5140, 0x7ffcd8231260, 0x442573, 0x7ffcd8231240, 0x40df9f, ...)
	/usr/local/go/src/runtime/sys_linux_amd64.s:587 +0x21
runtime.futexsleep(0x42a5b08, 0x3ed400000000, 0xffffffffffffffff)
	/usr/local/go/src/runtime/os_linux.go:45 +0x46
runtime.notesleep(0x42a5b08)
	/usr/local/go/src/runtime/lock_futex.go:159 +0x9f
runtime.stopm()
	/usr/local/go/src/runtime/proc.go:1924 +0xc5
runtime.exitsyscall0(0xc006280c00)
	/usr/local/go/src/runtime/proc.go:3415 +0x11d
runtime.mcall(0x0)
	/usr/local/go/src/runtime/asm_amd64.s:318 +0x5b

...

fs     0x0
gs     0x0
timeout: the monitored command dumped core
```

**Comment by lunny** [comment]
It seems it's timeout. unrelated with your PR.

File: models/commit_status.go [hunk 1]
```diff
@@ -21,12 +21,11 @@ import (
 
 // CommitStatus holds a single Status of a single Commit
 type CommitStatus struct {
-	ID          int64                 `xorm:"pk autoincr"`
-	Index       int64                 `xorm:"INDEX UNIQUE(repo_sha_index)"`
-	RepoID      int64                 `xorm:"INDEX UNIQUE(repo_sha_index)"`
+	ID          int64                 `xorm:"pk autoincr INDEX UNIQUE(repo_sha_id)"`
```
Comment (lafriks) [inline models/commit_status.go:24]: There is no point in adding Id to unique index as all records will be unique anyway, PK already guarantees that. It needs to be rechecked why unique index was set. (Maybe context needs to be added to that uq?)

**Comment by bobemoe** [comment]
I figured the original usage was to aid the `ORDER BY` which cant use the index on `id` alone if there is a `WHERE`, in which case the index needs to be multipart on `repo_id,sha,id` as the select will look like ```select `id`,`index`,`sha`,`state` from `commit_status` where `repo_id`=1 and `sha`='f0e325d9a81450c3f5a81249d19149a74d5e85c9' order by `index`;``` and I just tried to preserve that. But I wasn't sure how to define the order of columns in the index which is important. currently they in wrong order.

Although that said, the small amount or records for a specific SHA probably wont be that expensive to order without an index.

If there is no other usage for that index, maybe it could be removed.
---

From mysql docs:

```
SELECT * FROM t1
  WHERE key_part1 = constant
  ORDER BY key_part2;
```
> In this query, key_part1 is constant, so all rows accessed through the index are in key_part2 order, and an index on (key_part1, key_part2) avoids sorting if the WHERE clause is selective enough to make an index range scan cheaper than a table scan. https://dev.mysql.com/doc/refman/5.7/en/order-by-optimization.html#order-by-index-use

**Comment by kaenganxt** [comment]
Firstly, thanks for your work!
Will this be further worked on?
The bug is a major problem for my setup as my CI tool starts one check when another has finished and half of the time, the "finished" state of the first build is not shown in Gitea and thus blocks merging.

**Comment by lunny** [comment]
I will continue the work based on #15599

**Pull request closed without merging**